GG/src/RichText/ImageBlock.cpp

```diff
@@ -0,0 +1,152 @@
+        } else {
+            // We don't have an image. Take a quarter of width to show an X.
+            Pt size (width, Y (Value (width) / 4));
+            Resize (size);
+            return size;
+        }
+    }
+
+    void ImageBlock::Render() {
+        if (m_graphic == 0) {
+            // Error: no image. Draw red x.
+            Pt ul = UpperLeft();
+            Pt lr = LowerRight();
+            Pt size = lr - ul;
+            ul.x += size.x / 2 - X (Value (size.y)) / 2;
+            lr.x -= size.x / 2 - X (Value (size.y)) / 2;
+            FlatX (ul, lr, CLR_RED);
+        }
+    }
+
+    // A factory for creating image blocks from tags.
+    class ImageBlockFactory: public RichText::IBlockControlFactory {
+        public:
+            ImageBlockFactory() :
+                m_root_path (".")
+            {}
+
+            //! Create a Text block from a plain text tag.
+            virtual BlockControl* CreateFromTag (const std::string& tag,
+                                                 const RichText::TAG_PARAMS& params,
+                                                 const std::string& content,
+                                                 const boost::shared_ptr<Font>& font,
+                                                 Clr color,
+                                                 Flags<TextFormat> format
+                                                ) {
+                // Get the path from the parameters.
+                std::string path = ExtractPath (params);
+
+                // Create a new image block, basing the path on the root path.
+                return new ImageBlock ( (m_root_path / FileDlg::StringToPath(path)).string() , X0, Y0, X1, Flags<WndFlag>());
+            }
+
+            // Sets the root of image search path.
+            void SetRootPath (const std::string& path) {
+                m_root_path = FileDlg::StringToPath(path);
+            }
+        private:
+            boost::filesystem::path m_root_path;
+
+            // Extracts the path from the given params.
+            static std::string ExtractPath (const RichText::TAG_PARAMS& params) {
+
+                // Find the src.
+                RichText::TAG_PARAMS::const_iterator src_param = params.find ("src");
+
+                // If src not found, error out.
+                if (src_param == params.end()) {
+                    return std::string ("ERR:invalid params: missing src.");
+                } else {
+                    // Return the path.
+                    return src_param->second;
+                }
+            }
+    };
+
+    // Register image block as the image tag handler.
+    static int dummy = RichText::RegisterDefaultBlock (ImageBlock::IMAGE_TAG, new ImageBlockFactory());
+
+    //! Set the root path from which to look for images with the factory.
+    bool ImageBlock::SetImagePath (RichText::IBlockControlFactory* factory, //!< The factory to set the path for. Should be an image block factory.
```

Comment:
Perhaps should abandon the string path and root path stuff and just have ImageBlock take a Texture parameter like StaticGraphic does...? I suppose that makes in-GG img tag handling problematic, though?